### PR TITLE
Rtbkit 632 unify injection pattern

### PR DESCRIPTION
**Pull request by fhnmor21**
This is a redesign to address the need to unify the injection pattern. There was lots of replicated code to load libraries/plugins. Every one of the 8 base classes repeated the same bunch of code to load them. Everything is now managed by a template (singleton) class called PluginInterface.

There are 8 placeholder functions (marked FIXME) that are there only to maintain backwards compatible interface, as RTB still uses them. If all is good with this implementation, then RTB can be updated to use the new interface and the placeholder functions could be removed.

Most of the changes are in the new classes and the 8 base classes, the other ~30 files were changed to adapt to the new interface.

Currently these changes are under test at DEV5.

NOTE: the 8 base classes mentioned are BidRequest, BidderInterface, WinCostModel, ExchangeConnector, AdServerConnector, and ExchangeSource (Bid/Win/Event)

### CODE REVIEW

File: rtbkit/common/plugin_table.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <boost/any.hpp>
+#include "jml/arch/spinlock.h"
+
+namespace RTBKIT {
+
+struct PluginTable
```
Comment (oktal) [inline rtbkit/common/plugin_table.h]: Ok so now you have one static instance of the PluginTable per Factory type ?
I think I'm ok with that, @EricRobert ?

Comment (EricRobert) [inline rtbkit/common/plugin_table.h]: Looks good to me :)

Comment (fhnmor21) [inline rtbkit/common/plugin_table.h]: can I merge this?

**Comment by RAttab** [comment]
I think we can merge this now unless someone has an objection?

**Comment by fhnmor21** [comment]
squash merged.
closing this request

**Comment by fhnmor21** [comment]
I did a squash merge, few minutes ago.

**Pull request closed without merging**